content/posts/2020/2020-02-19-dask-clusters-overview.md

```diff
@@ -0,0 +1,337 @@
+
+Lastly the Python environment in the Python session where you create your `Client` must match the Python environment where the workers are running. This is because Dask uses [cloudpickle](https://github.com/cloudpipe/cloudpickle) to serialize objects and send them to workers and to retrieve results. Therefore package versions must match in both locations.
+
+We will need to bear these requirements in mind as we discuss the different platforms that folks generally want to run Dask on.
+
+## Cluster types
+
+There are two "types" of clusters that I tend to see folks running. Fixed clusters and ephemeral clusters.
+
+### Fixed clusters
+
+One common way of setting up a cluster is to run the scheduler and worker commands as described above, but leave them running indefinitely. For the purpose of this article I'll refer to this as a "fixed cluster". You may use something like [systemd](https://www.freedesktop.org/wiki/Software/systemd/) or [supervisord](http://supervisord.org/) to manage the processes and ensure they are always running on the machines. The Dask cluster can then be treated as a service.
+
+In this paradigm once a cluster is set up folks may start their Python session, connect their client to this existing cluster, do some work and disconnect again. They might later come back to that cluster and run further work. The cluster will sit idle in the meantime.
+
+It is also common in this paradigm for multiple users to share this single cluster, however this is not recommended as the Dask scheduler does not manage users or clients separately and work will be executed on a first come first served bases. Therefore we recommend that users use a cluster one at a time.
+
+### Ephemeral clusters
+
+An ephemeral cluster is one which only exists for the duration of the work. In this case a user may SSH onto the machines, run the commands to set up the cluster, connect a client and perform work, then disconnect and exit the Dask processes. A basic way of doing this would be to create a bash script which calls `ssh` and sets up the cluster. You would run this script in the background while performing your work and then kill it once you are done. We will cover other implementations of this in the coming sections.
+
+Ephemeral clusters allow you to leverage a bunch of machines but free them up again when you are done. This is especially useful when you are using a system like a cloud service or a batch scheduler where you have limited credits, or are charged for provisioned resources.
+
+## Adaptivity
+
+Ephemeral clusters are also generally easier to scale as you will likely have an automated mechanism for starting workers. The Dask scheduler maintains an estimate of how long it expects the outstanding work to take to complete. If the scheduler has a mechanism for starting and stopping workers then it will scale up the workers in an attempt to complete all outstanding work within 5 seconds. This is referred to as adaptive mode.
```

Comment:
```suggestion
Ephemeral clusters are also generally easier to scale as you will likely have an automated mechanism for starting workers. The Dask scheduler maintains an estimate of how long it expects the outstanding work will take to complete. If the scheduler has a mechanism for starting and stopping workers then it will scale up the workers in an attempt to complete all outstanding work within 5 seconds. This is referred to as adaptive mode.
```